content/posts/2023-11-20-tools-for-learning-llvm-tablegen.md

```diff
@@ -0,0 +1,433 @@
+ int size = 8;
+ string alias = "";
+}
+def X29 {       // Register
+ int size = 8;
+ string alias = "frame pointer";
+}
+```
+
+This is the intermediate representation of the TableGen compiler, akin to LLVM's
+LLVM IR.
+
+Unlike LLVM, TableGen’s “targets” are not instructions for CPUs but
+instead whatever output format the selected “backend” produces. For example
+there is one that generates C++ code that implements
+[searching](https://godbolt.org/z/5c696j1f9) of data tables.
+
+You might take your register definitions and produce C++ code to initialise them
+in some kind of bootloader. Perhaps you also document it and produce a diagram
+to illustrate it. With enough backends, you could do all that from the same
+TableGen source code.
+
+You would write these backends either in C++ within the TableGen compiler,
+or you can use the compiler’s [JSON output](https://godbolt.org/z/vre845e77)
+(`--dump-json`) and write them in any language with a JSON parser (e.g.
+[Python](https://github.com/llvm/llvm-project/blob/main/llvm/utils/TableGen/jupyter/sql_query_backend.ipynb)).
+
+# There is TableGen and There Are Things Built With TableGen
+
+This is more a mindset than a tool. It’s summed up best by a quote from the
+[documentation](https://llvm.org/docs/TableGen/index.html#tablegen-deficiencies):
+
+> Despite being very generic, TableGen has some deficiencies that have been
+> pointed out numerous times. The common theme is that, while TableGen allows
+> you to build domain specific languages, the final languages that you create
+> lack the power of other DSLs, which in turn increase considerably the size and
+> complexity of TableGen files.
+>
+> At the same time, TableGen allows you to create virtually any meaning of the
+> basic concepts via custom-made backends, which can pervert the original design
+> and make it very hard for newcomers to understand the evil TableGen file.”
+
+This means that you'll be tackling the language itself, and the things built
+with it. Which are often more complicated than the language.
+
+It’s like learning C++ and struggling to use [Boost](https://www.boost.org/).
+Someone might say to you “Boost isn’t required, why not remove it and save
+yourself the hassle?”. As someone new to C++, you might not be aware of the
+boundary between the two of them.
+
+Of course this doesn't help you too much if the project you want to contribute
+to uses Boost. You're stuck dealing with both. In LLVM terms, the TableGen
+language and the backends that consume it are a package deal.
+
+I bring this up so that you can draw a distinction between not understanding
+one or the other. Knowing which one is confusing you is a big advantage
+to finding help.
+
+For any given task there's probably one or two "things built with" you need to
+understand and even then not entirely.
+
+Don't think that the TableGen journey (for lack of a less grand term) ends with
+understanding the many many uses of it. If you could get there it would be an
```

Comment:
ends with understanding its many uses.

(try to not to end a sentence with a preposition, for example 'it'.

Reply:
Done.

Reply:
And I will fix a few others I found.